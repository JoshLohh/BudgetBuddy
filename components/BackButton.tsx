import React from 'react';
import { Pressable } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useRouter } from 'expo-router';

export default function BackButton({ color = 'black', size = 24, style = {} }) {
  const router = useRouter();
  return (
    <Pressable onPress={() => router.back()} style={[{ padding: 10 }, style]}>
      <Ionicons name="arrow-back" size={size} color={color} />
    </Pressable>
  );
}
