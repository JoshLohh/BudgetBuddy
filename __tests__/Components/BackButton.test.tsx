import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import BackButton from '../../components/BackButton';

// Mock expo-router's useRouter and back()
jest.mock('expo-router', () => {
  const backMock = jest.fn();
  return {
    useRouter: () => ({
      back: backMock,
    }),
    __esModule: true,
    _backMock: backMock,
  };
});

// âœ… Mock Ionicons with safe scoped factory using createElement
jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  return {
    Ionicons: (props: any) => React.createElement('icon', props, null),
  };
});

const getBackMock = () => require('expo-router')._backMock;

describe('BackButton Component', () => {
  beforeEach(() => {
    getBackMock().mockClear();
  });

  it('renders correctly with default props', () => {
    const { getByRole } = render(<BackButton />);
    const pressable = getByRole('button');
    expect(pressable).toBeTruthy();
  });

  it('renders Ionicons with correct props when overridden', () => {
    const { getByTestId } = render(<BackButton color="red" size={30} />);
    const icon = getByTestId('back-icon');
    expect(icon.props.color).toBe('red');
    expect(icon.props.size).toBe(30);
  });

  it('merges custom and default styles', () => {
    const { getByRole } = render(<BackButton style={{ margin: 5 }} />);
    const { style } = getByRole('button').props;
    expect(style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ padding: 10 }),
        expect.objectContaining({ margin: 5 }),
      ])
    );
  });

  it('calls router.back when pressed', () => {
    const { getByRole } = render(<BackButton />);
    fireEvent.press(getByRole('button'));
    expect(getBackMock()).toHaveBeenCalledTimes(1);
  });

  it('matches the UI snapshot', () => {
    const { toJSON } = render(<BackButton color="blue" size={28} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
