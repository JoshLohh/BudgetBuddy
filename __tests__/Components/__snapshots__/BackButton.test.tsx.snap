// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton Component matches the UI snapshot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "padding": 10,
      },
      {},
    ]
  }
>
  <icon
    color="blue"
    name="arrow-back"
    size={28}
    testID="back-icon"
  />
</View>
`;
